### ISSUE TITLE
Finetuning with LoRA and video training data.

### ISSUE THREAD

**Issue by tungnt55** [2U1/Qwen2-VL-Finetune#170]
I use the following bash script for finetuning model using LoRA. The loss went down to almost zero, however, when I tried to evaluate the model using the exact data set for training, accuracy is only about 60-70%.
Can someone help me pinpointing the problem with settings below?

```bash
#!/bin/bash

# You can use 2B instead of 7B
# MODEL_NAME="Qwen/Qwen2-VL-7B-Instruct"
# MODEL_NAME="Qwen/Qwen2-VL-2B-Instruct"
MODEL_NAME="Qwen/Qwen2.5-VL-3B-Instruct"
# MODEL_NAME="Qwen/Qwen2.5-VL-7B-Instruct"

export PYTHONPATH=src:$PYTHONPATH

GLOBAL_BATCH_SIZE=128
BATCH_PER_DEVICE=4
NUM_DEVICES=8
GRAD_ACCUM_STEPS=$((GLOBAL_BATCH_SIZE / (BATCH_PER_DEVICE * NUM_DEVICES)))

# If your dataset is mixed with images and videos, you need to use zero2.
CUDA_VISIBLE_DEVICES=4,5,6,7 deepspeed src/train/train_sft.py \
    --use_liger True \
    --deepspeed scripts/zero3_offload.json \
    --model_id $MODEL_NAME \
    --data_path my_data.json \
    --image_folder /path/to/your/image/folder \
    --remove_unused_columns False \
    --freeze_vision_tower False \
    --freeze_llm False \
    --freeze_merger False \
    --bf16 True \
    --fp16 False \
    --disable_flash_attn2 False \
    --output_dir output/finetuned_test_4fps_lora \
    --num_train_epochs 100 \
    --per_device_train_batch_size $BATCH_PER_DEVICE \
    --gradient_accumulation_steps $GRAD_ACCUM_STEPS \
    --video_max_pixels $((1280 * 720)) \
    --fps 4 \
    --learning_rate 1e-5 \
    --merger_lr 1e-5 \
    --vision_lr 2e-6 \
    --weight_decay 0.1 \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --gradient_checkpointing True \
    --report_to tensorboard \
    --lazy_preprocess True \
    --save_strategy "steps" \
    --save_steps 5 \
    --save_total_limit 10 \
    --dataloader_num_workers 4
```

**Comment by 2U1** [comment]
Are you measuring the exact token answer or the perplexity?
Also, the model could have problem like "catastrophic forgetting".

**Comment by 2U1** [comment]
Did you merge the lora weights? 
Or it could be a issue that updates to the lora weight was insufficient. You could increase the rank and the alpha

**Comment by tungnt55** [comment]
I have tried increase rank and alpha to 96. Results are still strange.
At 100 epoch, the loss is 0.01 and the accuracy is 70%. 
At 250 epoch, the loss is 0.0 and accuracy is 66%.
I was using the unmerged weights in these tests.

**Comment by tungnt55** [comment]
I have double-checked and confirm that the prompt or text processing is not the issue. Also, under the same setting, training without LoRA works OK, the trained model reach 100% accuracy when their train loss is 0.
I also tried again, this time train with 2 epochs.
Epoch 1, loss: 0.36, accuracy: 33%
Epoch 2, loss: 0.16, accuracy: 30%.

**Comment by 2U1** [comment]
You should freeze the merger too.

https://github.com/2U1/Qwen2-VL-Finetune/blob/d5ebc1a417b1d1bf1da24ba8c20e8406356cdcb2/scripts/finetune_lora_vision.sh#L17

**Comment by tungnt55** [comment]
I have tried again with --freeze_merger=True but it didn't change.
Below is the code that I use for loading the unmerged weights. Could you take a quick look and see if there is any issue with it?

```python
model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    base_model_path,
    low_cpu_mem_usage=True,
    torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
    local_files_only=True
)
model.to(device)

non_lora_path = os.path.join(model_path, 'non_lora_state_dict.bin')
non_lora_trainables = torch.load(non_lora_path, map_location='cpu')


non_lora_trainables = {
    (k[11:] if k.startswith('base_model.') else k): v
    for k, v in non_lora_trainables.items()
}
if any(k.startswith('model.model.') for k in non_lora_trainables):
    non_lora_trainables = {
        (k[6:] if k.startswith('model.') else k): v
        for k, v in non_lora_trainables.items()
    }

model.load_state_dict(non_lora_trainables, strict=False)

model = PeftModel.from_pretrained(model, model_path,is_trainable=False)

model.eval()
```

**Comment by 2U1** [comment]
```
res = model.load_state_dict(non_lora_trainables, strict=False)
print(f"[non-lora] missing: {len(res.missing_keys)}, unexpected: {len(res.unexpected_keys)}")

print("example missing (merger):", [k for k in res.missing_keys if "visual.merger" in k][:10])
```
check if the prefixes are correct.


```
def strip_prefixes(state, prefixes=("base_model.", "model.", "model.model.")):
    out = {}
    for k, v in state.items():
        kk = k
        for p in prefixes:
            if kk.startswith(p):
                kk = kk[len(p):]
        out[kk] = v
    return out

non_lora_trainables = strip_prefixes(non_lora_trainables)
res = model.load_state_dict(non_lora_trainables, strict=False)
print(f"[non-lora] missing: {len(res.missing_keys)}, unexpected: {len(res.unexpected_keys)}")
```
This is the safe one.

**Comment by tungnt55** [comment]
Thank you for sending me the code to check it.
I tried printing out res.missking_keys and res.unexpected_keys here:
[lora_load_state_dict.log](https://github.com/user-attachments/files/21679758/lora_load_state_dict.log)
I modified the function strip_prefixes() a little bit to fix the issue of missing keys, after fixing, only 'lm_head.weight' remains missing. 
However, even after this fix, the accuracy on same dataset remains low when the loss is also very low (close to zero) in training.

**Comment by tungnt55** [comment]
Inside sft_trainer.py, I added new code to evaluate with two different methods. Below is the result.
Calling model(**input) gets perfect accuracy, which is inline with low loss, while calling model.generate(**input) still gets some errrors.
```
{'loss': 0.0055, 'grad_norm': 0.27690991168849555, 'learning_rate': 1.3128826348184887e-05, 'epoch': 84.75}
{'loss': 0.002, 'grad_norm': 0.27690991168849555, 'learning_rate': 1.1570450926997655e-05, 'epoch': 85.0}

Evaluate with model(**inputs)...                                                                                                                             
Target text: ['high', 'high', 'low', 'low']
Output predicted text: ['high', 'high', 'low', 'low']
Target text: ['high', 'low', 'low', 'high']
Output predicted text: ['high', 'low', 'low', 'high']
Target text: ['medium', 'low', 'high', 'medium']
Output predicted text: ['medium', 'low', 'high', 'medium']
Target text: ['high', 'low', 'high', 'high']
Output predicted text: ['high', 'low', 'high', 'high']
Target text: ['low', 'medium', 'high', 'low']
Output predicted text: ['low', 'medium', 'high', 'low']
Target text: ['high', 'low', 'high', 'medium']
Output predicted text: ['high', 'low', 'high', 'medium']
Target text: ['low', 'medium', 'high', 'high']
Output predicted text: ['low', 'medium', 'high', 'high']
Target text: ['high', 'low']
Output predicted text: ['high', 'low']


Evaluate with model.generate(**inputs)...
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "medium"}'] <-- INCORRECT
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: high
Output predicted text: ['{"result": "medium"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: low
Output predicted text: ['{"result": "medium"}'] <-- INCORRECT
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: medium
Output predicted text: ['{"result": "medium"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: medium
Output predicted text: ['{"result": "low"}'] <-- INCORRECT
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: medium
Output predicted text: ['{"result": "low"}'] <-- INCORRECT
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: medium
Output predicted text: ['{"result": "medium"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "medium"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "low"}']
Target text: medium
Output predicted text: ['{"result": "low"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: high
Output predicted text: ['{"result": "high"}']
Target text: low
Output predicted text: ['{"result": "low"}']
```

**Comment by 2U1** [comment]
model.generate will use some other generation settings such as top_k or top_p. 
You should check those.
